SCClassLibrary/Common/Control/Server.sc

```diff
@@ -884,13 +885,36 @@ Server {
 				this.quit;
 				this.boot;
 			}, {
-				this.bootServerApp({
-					if(startAliveThread) { statusWatcher.startAliveThread }
-				})
+				this.waitForPidRelease {
+					this.bootServerApp({
+						if(startAliveThread) { statusWatcher.startAliveThread }
+					})
+				};
 			}, 0.25);
 		}
 	}
 
+	waitForPidRelease { |onComplete, onFailure, timeout = 1|
+		var bootStart;
+		if (this.inProcess or: { this.isLocal.not }) {
+			onComplete.value;
+			^this
+		};
+		// FIXME: quick and dirty fix for supernova reboot hang on osx:
+		// if we have just quit before running server.boot,
+		// we wait until server process really ends and sets its pid to nil
+		forkIfNeeded {
+			bootStart = Main.elapsedTime;
+			while {
+				pid.notNil and: {
+					Main.elapsedTime - bootStart < timeout
+				}
+			} {
+				0.05.wait
+			};
```

Comment:
please rewrite this using a `Condition` - waiting in a loop is inefficient and we have better idioms for this purpose

Reply:
I'm not sure: do we have a callback somewhere that would make this possible? There would have to be a mechanism that unhangs the condition when the pid becomes nil. As long as we don't have that, this loop is the only possible way.

Reply:
A possible way to implement it is that any code that changes the pid can `signal` a Condition that is `wait`ed on here. The Condition will have to be at class scope. `pid == nil` can even be used as the test function for the Condition.

Reply:
If we do this, the condition has to be at object scope, otherwise booting servers will block each other (but this is probably what you meant, anyhow).

Unless there are other uses for a condition at that level, I'd prefer to keep the method self-contained. Polling is not bad practice as such at all, it just depends on the case. But I'm fine either way if it is elegant.

Reply:
I don't think efficiency is an issue here - waiting until something has finished properly so you can continue is very cheap. As Julian points out, the polling solution here stays completely local; also, it is just a single local poll/wait thread, so there is no possible confusion of order of execution of waiting tasks. 
 IIRC, Condition.hangWithTimeout (which would be needed here) ran into problems that are still not resolved yet, right? 
To me, polling seems the right choice here, and conditions would make it more complex than necessary.

Reply:
This would a case for whenTrueWithin, as proposed in #3850.

Reply:
Do whatever you want. This is not just about efficiency, but this code is going to be rewritten soon anyway.

Reply:
yes, eg. when I get around to redoing the ServerProcess refactor :-) 
which is hopefully soon